@here/harp-mapview/lib/BoundsGenerator.ts

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (C) 2017-2020 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import {
+    EarthConstants,
+    GeoCoordinates,
+    mercatorTilingScheme,
+    ProjectionType,
+    TileKey
+} from "@here/harp-geoutils";
+import { Frustum, Line3, Matrix4, Plane, Vector2, Vector3 } from "three";
+
+import { TileGeometryCreator } from "./geometry/TileGeometryCreator";
+import { MapView } from "./MapView";
+import { MapViewUtils } from "./Utils";
+
+export class BoundsGenerator {
+    //TODO: check what we really need here
+    constructor(private m_mapView: MapView) {}
+
+    generate(): GeoCoordinates[] {
+        //TODO: support spherical projection
+
+        //!!!!!!!ALTITUDE IS NOT TAKEN INTO ACCOUNT!!!!!!!!!
+        const geoPolygon: GeoCoordinates[] = [];
+
+        //CASE A: FLAT PROJECTION
+        if (this.m_mapView.projection.type === ProjectionType.Planar) {
+            // 1.) Raycast into all for corners of the canvas
+            //     => if an intersection is found, add it to the polygon
+            this.raycastCanvasCorners(geoPolygon);
+
+            // => All 4 corners found an intersection, therefore the screen is covered with the map
+            // and the polygon complete
+            if (geoPolygon.length === 4) {
+                return geoPolygon;
+            }
+
+            //2.) Raycast into the two corners of the horizon cutting the canvas sides
+            //    => if intersection is found, add it to the polygon
+            this.rayCastOnHorizonCorners(geoPolygon);
+
+            //Setup the frustum for further checks
+            const frustum = new Frustum().setFromProjectionMatrix(
```

Comment:
Can we not reuse the frustum that is also used for the visible tile intersection?

Reply:
probably, it should be the same, I ll check how I can expose it for reuse

Reply:
The FrustumIntersection is nested inside the VisibleTileSet, and creates the frustum as private property. The only place this and BoundsGenerate have in common right now is the MapView. But i think adding the frustum to MapView would just overload MapView even more. Would be nice to have it directly on the camera... but we are using THREE.Camera, .. ideas are appreciated, but for now I would keep it here, it should not be to expensive to have this instance here.

Reply:
It would just be a private property in `MapView` that is passed to the ctor of `VisibleTileSet` and to the ctor of `BoundGenerator`.
I agree the computation is not that expensive but it's at least fishy to have 2 frustums. If some magic is added to the computation of one of them we end up using two different frustums.
By just writing that sentence I realized it is actually the case. Check:
https://github.com/heremaps/harp.gl/blob/master/@here/harp-mapview/lib/FrustumIntersection.ts#L117

Reply:
I had a deeper look at the Frustum being used in FrustumIntersection.ts, there is two reasons, why I think we should/can not share the two:
1. In some cases (terrain) the FrustumIntersection is using a frustum that does not match the visible frustum of the actual camera, to calculate the needed tiles. In the case of the BoundsGeneration, this frustum would not be correct, as the bounds calculation  needs the actually visible frustum.
2. The frustum changes with the camera, we already pass the camera to the BoundsGeneration which is enough to retrieve the Frustum, the camera is controlled and updated by the MapView in our case, but could basically be any camera that does not have to be related to MapView. If we create another dependency on a Frustum, the updating of this frustum depending on the given camera would need to be taken care of, which is redundant and error prone